### ISSUE TITLE
When SQLFluff lint folder, the exit code depends on the last file

### ISSUE THREAD

**Issue by kang8** [sqlfluff/sqlfluff#5673]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When I upgraded to SQLFluff 3.0.0, I found that the `sqlfluff lint .` command executed in CI did not fail, even though there were files that found issues.

After some simple debugging, I found that when linting a folder, the exit code depends on the last file: If the last file found the issue, the command result is `FAIL`; If not issue is found in the last file, the command result is `SUCCESS`.

### Expected Behaviour

When running the `sqlfluff lint .` command, if any issues are found with the file, exit code return 1 or 2. 

Just like it says in the [document](https://github.com/sqlfluff/sqlfluff/blob/85c85bf0dedb9608e300e1214fa0e3a0b7228f1d/docs/source/production.rst?plain=1#L66-L81).

```
The `exit code`_ provided by SQLFluff when run as a command line utility is
designed to assist usefulness in deployment pipelines. If no violations
are found then the `exit code`_ will be 0. If violations are found then
a non-zero code will be returned which can be interrogated to find out
more.

- An error code of ``0`` means *operation success*, *no issues found*.
- An error code of ``1`` means *operation success*, *issues found*. For
  example this might mean that a linting issue was found, or that one file
  could not be parsed.
- An error code of ``2`` means an error occurred and the operation could
  not be completed. For example a configuration issue or an internal error
  within SQLFluff.
```

### Observed Behaviour

When running the `sqlfluff lint .` command, the exit code depends on the last file:

* When the last file not issues found, exit code is 0.
* When the last file issues found, exit code is 1.

> PS: At the same time, I tested it in 2.3.5 and did not find this issue.

### How to reproduce

SQLFluff is executed in alphabetical order in the directory, so creating two files to make one of them fail can verify this conclusion.

## It is not the last file that failed, the result is `SUCCESS`

1. Create an empty directory

```bash
mkdir sqlfluff-playgrand
```

2. Create the two sql files: `1.sql`, `2.sql`. `1.sql` will fail and `2.sql` will pass.

The content of `1.sql` is as follows:
```sql
select
    column
from table
```

The content of `2.sql` is as follows:
```sql
select column
from table
```

3. running the `sqlfluff lint . --dialect ansi` in the directory, will return `SUCCESS` results

```bash
sqlfluff lint . --dialect ansi
== [path/test-playgrand/1.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target.
                       | [layout.select_targets]
All Finished 📜 🎉! 
```

<details>
<summary> The output of `sqlfluff lint . --dialect ansi -vv` is </summary>

```bash
sqlfluff lint . --dialect ansi -vv
==== sqlfluff ====
sqlfluff:                3.0.0 python:                 3.12.2
implementation:        cpython verbosity:                   2
dialect:                  ansi templater:               jinja
rules:                                all
== Raw Config:
core:                                   
    dialect:            ansi                
    disable_noqa:       False               
    encoding:           autodetect          
    fix_even_unparsable:False               
    ignore:             []                  
    ignore_templated_areas:True                
    large_file_skip_byte_limit:20000               
    large_file_skip_char_limit:0                   
    max_line_length:    80                  
    nocolor:            False               
    output_line_length: 80                  
    processes:          1                   
    rules:              all                 
    runaway_limit:      10                  
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja               
    verbose:            2                   
    warn_unused_ignores:False               
    warnings:           []                  
indentation:                            
    allow_implicit_indents:False               
    ignore_comment_lines:False               
    indent_unit:        space               
    indented_ctes:      False               
    indented_joins:     False               
    indented_on_contents:True                
    indented_then:      True                
    indented_then_contents:True                
    indented_using_on:  True                
    skip_indentation_in:script_content      
    tab_space_size:     4                   
    template_blocks_indent:True                
    trailing_comments:  before              
layout:                                 
    type:                                   
        array_accessor:                         
            spacing_before:     touch:inline        
        array_type:                             
            spacing_within:     touch:inline        
        assignment_operator:                    
            line_position:      leading             
            spacing_within:     touch               
        binary_operator:                        
            line_position:      leading             
            spacing_within:     touch               
        bracketed_arguments:                    
            spacing_before:     touch:inline        
        casting_operator:                       
            spacing_after:      touch:inline        
            spacing_before:     touch               
        colon:                                  
            spacing_before:     touch               
        colon_delimiter:                        
            spacing_after:      touch               
            spacing_before:     touch               
        comma:                                  
            line_position:      trailing            
            spacing_before:     touch               
        comment:                                
            spacing_after:      any                 
            spacing_before:     any                 
        common_table_expression:                    
            spacing_within:     single:inline       
        comparison_operator:                    
            line_position:      leading             
            spacing_within:     touch               
        dot:                                    
            spacing_after:      touch               
            spacing_before:     touch               
        end_angle_bracket:                      
            spacing_before:     touch               
        end_bracket:                            
            spacing_before:     touch               
        end_of_file:                            
            spacing_before:     touch               
        end_square_bracket:                     
            spacing_before:     touch               
        from_clause:                            
            line_position:      alone               
        function_name:                          
            spacing_after:      touch:inline        
            spacing_within:     touch:inline        
        groupby_clause:                         
            line_position:      alone               
        having_clause:                          
            line_position:      alone               
        join_clause:                            
            line_position:      alone               
        limit_clause:                           
            line_position:      alone               
        numeric_literal:                        
            spacing_within:     touch:inline        
        object_reference:                       
            spacing_within:     touch:inline        
        orderby_clause:                         
            line_position:      leading             
        path_segment:                           
            spacing_within:     touch               
        pattern_expression:                     
            spacing_within:     any                 
        placeholder:                            
            spacing_after:      any                 
            spacing_before:     any                 
        select_clause:                          
            line_position:      alone               
        semi_structured_expression:                    
            spacing_before:     touch:inline        
            spacing_within:     touch:inline        
        set_operator:                           
            line_position:      alone:strict        
        sign_indicator:                         
            spacing_after:      touch:inline        
        sized_array_type:                       
            spacing_within:     touch               
        slice:                                  
            spacing_after:      touch               
            spacing_before:     touch               
        sql_conf_option:                        
            spacing_within:     touch               
        sqlcmd_operator:                        
            spacing_before:     touch               
        start_angle_bracket:                    
            spacing_after:      touch               
        start_bracket:                          
            spacing_after:      touch               
        start_square_bracket:                    
            spacing_after:      touch               
        statement_terminator:                    
            line_position:      trailing            
            spacing_before:     touch               
        struct_type:                            
            spacing_within:     touch:inline        
        template_loop:                          
            spacing_after:      any                 
            spacing_before:     any                 
        tilde:                                  
            spacing_after:      touch:inline        
        typed_array_literal:                    
            spacing_within:     touch               
        typed_struct_literal:                    
            spacing_within:     touch               
        where_clause:                           
            line_position:      alone               
rules:                                  
    allow_scalar:       True                
    single_table_references:consistent          
    unquoted_identifiers_policy:all                 
    aliasing.column:                        
        aliasing:           explicit            
    aliasing.forbid:                        
        force_enable:       False               
    aliasing.length:                        
    aliasing.table:                         
        aliasing:           explicit            
    ambiguous.column_references:                    
        group_by_and_order_by_style:consistent          
    ambiguous.join:                         
        fully_qualify_join_types:inner               
    capitalisation.functions:                    
        extended_capitalisation_policy:consistent          
    capitalisation.identifiers:                    
        extended_capitalisation_policy:consistent          
    capitalisation.keywords:                    
        capitalisation_policy:consistent          
    capitalisation.literals:                    
        capitalisation_policy:consistent          
    capitalisation.types:                    
        extended_capitalisation_policy:consistent          
    convention.blocked_words:                    
        match_source:       False               
    convention.casting_style:                    
        preferred_type_casting_style:consistent          
    convention.count_rows:                    
        prefer_count_0:     False               
        prefer_count_1:     False               
    convention.not_equal:                    
        preferred_not_equal_style:consistent          
    convention.quoted_literals:                    
        force_enable:       False               
        preferred_quoted_literal_style:consistent          
    convention.select_trailing_comma:                    
        select_clause_trailing_comma:forbid              
    convention.terminator:                    
        multiline_newline:  False               
        require_final_semicolon:False               
    layout.long_lines:                      
        ignore_comment_clauses:False               
        ignore_comment_lines:False               
    layout.select_targets:                    
        wildcard_policy:    single              
    references.consistent:                    
        force_enable:       False               
    references.from:                        
        force_enable:       False               
    references.keywords:                    
        unquoted_identifiers_policy:aliases             
    references.qualification:                    
    references.quoting:                     
        force_enable:       False               
        prefer_quoted_identifiers:False               
        prefer_quoted_keywords:False               
    references.special_chars:                    
        allow_space_in_identifier:False               
        quoted_identifiers_policy:all                 
        unquoted_identifiers_policy:all                 
    structure.join_condition_order:                    
        preferred_first_table_in_join_clause:earlier             
    structure.subquery:                     
        forbid_subquery_in: join                
templater:                              
    unwrap_wrapped_queries:True                
    jinja:                                  
        apply_dbt_builtins: True                

==== readout ====

== [path/test-playgrand/1.sql] LINTING (AL01, AL02, AL03, AL04, AL05, AL06, AL07, AL08, AL09, AM01, AM02, AM03, AM04, AM05, AM06, AM07, CP01, CP02, CP03, CP04, CP05, CV01, CV02, CV03, CV04, CV05, CV06, CV07, CV08, CV09, CV10, CV11, JJ01, LT01, LT02, LT03, LT04, LT05, LT06, LT07, LT08, LT09, LT10, LT11, LT12, LT13, RF01, RF02, RF03, RF04, RF05, RF06, ST01, ST02, ST03, ST04, ST05, ST06, ST07, ST08, ST09, TQ01)
== [path/test-playgrand/1.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
== [path/test-playgrand/2.sql] LINTING (AL01, AL02, AL03, AL04, AL05, AL06, AL07, AL08, AL09, AM01, AM02, AM03, AM04, AM05, AM06, AM07, CP01, CP02, CP03, CP04, CP05, CV01, CV02, CV03, CV04, CV05, CV06, CV07, CV08, CV09, CV10, CV11, JJ01, LT01, LT02, LT03, LT04, LT05, LT06, LT07, LT08, LT09, LT10, LT11, LT12, LT13, RF01, RF02, RF03, RF04, RF05, RF06, ST01, ST02, ST03, ST04, ST05, ST06, ST07, ST08, ST09, TQ01)
== [path/test-playgrand/2.sql] PASS
==== summary ====
files:             2 violations:        0
clean files:       1 unclean files:     1
avg per file:   0.00 unclean rate:    50%
status:         PASS
All Finished!
```

</details>

And `echo $?` will output 0

```bash
echo $?
0
```

## The last file failed, the result is `FAIL`

1. Create an empty directory

```bash
mkdir sqlfluff-playgrand
```

2. Create the two sql files: `1.sql`, `2.sql`. `1.sql` will pass and `2.sql` will fail.

The content of `1.sql` is as follows:
```sql
select column
from table
```

The content of `2.sql` is as follows:
```sql
select
    column
from table
```

3. running the `sqlfluff lint . --dialect ansi` in the directory, will return `FAIL` results

```bash
sqlfluff lint . --dialect ansi
== [path/test-playgrand/2.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target.
                       | [layout.select_targets]
All Finished 📜 🎉! 
```

<details>
<summary>The output of `sqlfluff lint . --dialect ansi -vv` is </summary>

```bash
sqlfluff lint . --dialect ansi -vv
==== sqlfluff ====
sqlfluff:                3.0.0 python:                 3.12.2
implementation:        cpython verbosity:                   2
dialect:                  ansi templater:               jinja
rules:                                all
== Raw Config:
core:                                   
    dialect:            ansi                
    disable_noqa:       False               
    encoding:           autodetect          
    fix_even_unparsable:False               
    ignore:             []                  
    ignore_templated_areas:True                
    large_file_skip_byte_limit:20000               
    large_file_skip_char_limit:0                   
    max_line_length:    80                  
    nocolor:            False               
    output_line_length: 80                  
    processes:          1                   
    rules:              all                 
    runaway_limit:      10                  
    sql_file_exts:      .sql,.sql.j2,.dml,.ddl
    templater:          jinja               
    verbose:            2                   
    warn_unused_ignores:False               
    warnings:           []                  
indentation:                            
    allow_implicit_indents:False               
    ignore_comment_lines:False               
    indent_unit:        space               
    indented_ctes:      False               
    indented_joins:     False               
    indented_on_contents:True                
    indented_then:      True                
    indented_then_contents:True                
    indented_using_on:  True                
    skip_indentation_in:script_content      
    tab_space_size:     4                   
    template_blocks_indent:True                
    trailing_comments:  before              
layout:                                 
    type:                                   
        array_accessor:                         
            spacing_before:     touch:inline        
        array_type:                             
            spacing_within:     touch:inline        
        assignment_operator:                    
            line_position:      leading             
            spacing_within:     touch               
        binary_operator:                        
            line_position:      leading             
            spacing_within:     touch               
        bracketed_arguments:                    
            spacing_before:     touch:inline        
        casting_operator:                       
            spacing_after:      touch:inline        
            spacing_before:     touch               
        colon:                                  
            spacing_before:     touch               
        colon_delimiter:                        
            spacing_after:      touch               
            spacing_before:     touch               
        comma:                                  
            line_position:      trailing            
            spacing_before:     touch               
        comment:                                
            spacing_after:      any                 
            spacing_before:     any                 
        common_table_expression:                    
            spacing_within:     single:inline       
        comparison_operator:                    
            line_position:      leading             
            spacing_within:     touch               
        dot:                                    
            spacing_after:      touch               
            spacing_before:     touch               
        end_angle_bracket:                      
            spacing_before:     touch               
        end_bracket:                            
            spacing_before:     touch               
        end_of_file:                            
            spacing_before:     touch               
        end_square_bracket:                     
            spacing_before:     touch               
        from_clause:                            
            line_position:      alone               
        function_name:                          
            spacing_after:      touch:inline        
            spacing_within:     touch:inline        
        groupby_clause:                         
            line_position:      alone               
        having_clause:                          
            line_position:      alone               
        join_clause:                            
            line_position:      alone               
        limit_clause:                           
            line_position:      alone               
        numeric_literal:                        
            spacing_within:     touch:inline        
        object_reference:                       
            spacing_within:     touch:inline        
        orderby_clause:                         
            line_position:      leading             
        path_segment:                           
            spacing_within:     touch               
        pattern_expression:                     
            spacing_within:     any                 
        placeholder:                            
            spacing_after:      any                 
            spacing_before:     any                 
        select_clause:                          
            line_position:      alone               
        semi_structured_expression:                    
            spacing_before:     touch:inline        
            spacing_within:     touch:inline        
        set_operator:                           
            line_position:      alone:strict        
        sign_indicator:                         
            spacing_after:      touch:inline        
        sized_array_type:                       
            spacing_within:     touch               
        slice:                                  
            spacing_after:      touch               
            spacing_before:     touch               
        sql_conf_option:                        
            spacing_within:     touch               
        sqlcmd_operator:                        
            spacing_before:     touch               
        start_angle_bracket:                    
            spacing_after:      touch               
        start_bracket:                          
            spacing_after:      touch               
        start_square_bracket:                    
            spacing_after:      touch               
        statement_terminator:                    
            line_position:      trailing            
            spacing_before:     touch               
        struct_type:                            
            spacing_within:     touch:inline        
        template_loop:                          
            spacing_after:      any                 
            spacing_before:     any                 
        tilde:                                  
            spacing_after:      touch:inline        
        typed_array_literal:                    
            spacing_within:     touch               
        typed_struct_literal:                    
            spacing_within:     touch               
        where_clause:                           
            line_position:      alone               
rules:                                  
    allow_scalar:       True                
    single_table_references:consistent          
    unquoted_identifiers_policy:all                 
    aliasing.column:                        
        aliasing:           explicit            
    aliasing.forbid:                        
        force_enable:       False               
    aliasing.length:                        
    aliasing.table:                         
        aliasing:           explicit            
    ambiguous.column_references:                    
        group_by_and_order_by_style:consistent          
    ambiguous.join:                         
        fully_qualify_join_types:inner               
    capitalisation.functions:                    
        extended_capitalisation_policy:consistent          
    capitalisation.identifiers:                    
        extended_capitalisation_policy:consistent          
    capitalisation.keywords:                    
        capitalisation_policy:consistent          
    capitalisation.literals:                    
        capitalisation_policy:consistent          
    capitalisation.types:                    
        extended_capitalisation_policy:consistent          
    convention.blocked_words:                    
        match_source:       False               
    convention.casting_style:                    
        preferred_type_casting_style:consistent          
    convention.count_rows:                    
        prefer_count_0:     False               
        prefer_count_1:     False               
    convention.not_equal:                    
        preferred_not_equal_style:consistent          
    convention.quoted_literals:                    
        force_enable:       False               
        preferred_quoted_literal_style:consistent          
    convention.select_trailing_comma:                    
        select_clause_trailing_comma:forbid              
    convention.terminator:                    
        multiline_newline:  False               
        require_final_semicolon:False               
    layout.long_lines:                      
        ignore_comment_clauses:False               
        ignore_comment_lines:False               
    layout.select_targets:                    
        wildcard_policy:    single              
    references.consistent:                    
        force_enable:       False               
    references.from:                        
        force_enable:       False               
    references.keywords:                    
        unquoted_identifiers_policy:aliases             
    references.qualification:                    
    references.quoting:                     
        force_enable:       False               
        prefer_quoted_identifiers:False               
        prefer_quoted_keywords:False               
    references.special_chars:                    
        allow_space_in_identifier:False               
        quoted_identifiers_policy:all                 
        unquoted_identifiers_policy:all                 
    structure.join_condition_order:                    
        preferred_first_table_in_join_clause:earlier             
    structure.subquery:                     
        forbid_subquery_in: join                
templater:                              
    unwrap_wrapped_queries:True                
    jinja:                                  
        apply_dbt_builtins: True                

==== readout ====

== [/Users/kang/playground/data/test-playgrand/1.sql] LINTING (AL01, AL02, AL03, AL04, AL05, AL06, AL07, AL08, AL09, AM01, AM02, AM03, AM04, AM05, AM06, AM07, CP01, CP02, CP03, CP04, CP05, CV01, CV02, CV03, CV04, CV05, CV06, CV07, CV08, CV09, CV10, CV11, JJ01, LT01, LT02, LT03, LT04, LT05, LT06, LT07, LT08, LT09, LT10, LT11, LT12, LT13, RF01, RF02, RF03, RF04, RF05, RF06, ST01, ST02, ST03, ST04, ST05, ST06, ST07, ST08, ST09, TQ01)
== [/Users/kang/playground/data/test-playgrand/1.sql] PASS
== [/Users/kang/playground/data/test-playgrand/2.sql] LINTING (AL01, AL02, AL03, AL04, AL05, AL06, AL07, AL08, AL09, AM01, AM02, AM03, AM04, AM05, AM06, AM07, CP01, CP02, CP03, CP04, CP05, CV01, CV02, CV03, CV04, CV05, CV06, CV07, CV08, CV09, CV10, CV11, JJ01, LT01, LT02, LT03, LT04, LT05, LT06, LT07, LT08, LT09, LT10, LT11, LT12, LT13, RF01, RF02, RF03, RF04, RF05, RF06, ST01, ST02, ST03, ST04, ST05, ST06, ST07, ST08, ST09, TQ01)
== [/Users/kang/playground/data/test-playgrand/2.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
==== summary ====
files:             2 violations:        1
clean files:       1 unclean files:     1
avg per file:   0.50 unclean rate:    50%
status:         FAIL
All Finished!
```
</details>

And `echo $?` will output 1

```bash
echo $?
1
```

### Dialect

I have tested it in both `ANSI` and `Snowflake`, and encountered the same issue. I believe other dialects will also encounter the same issue.

### Version

3.0.0

### Configuration

For this case, it does not need `.sqlfluff` and `sqlfluffignore` file.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
This is definitely a bug - thanks for reporting it. I'm investigating now.

**Comment by alanmcruickshank** [comment]
Found the bug - it's a really tedious typo 🤦 .

The fix is very easy though. Will get a hotfix out soon.

**Comment by alanmcruickshank** [comment]
3.0.1 release is out now 👍

**Comment by kang8** [comment]
Thank you for the quick fix
